Replace `com.jcraft:jsch` with `com.github.mwiede:jsch`

aSemy opened the pull request:
<!--- The issue this PR addresses -->

Fixes https://github.com/gradle/gradle/issues/19028

(second attempt, after https://github.com/gradle/gradle/pull/19060)

### Context

Gradle declares a dependency on `com.jcraft:jsch`. This library is abandoned and does not support several important features, like modern features like ed25519 keys.

[`com.github.mwiede:jsch`](https://github.com/mwiede/jsch) is a maintained, drop-in replacement. 

See also 

* https://stackoverflow.com/q/56062772/4161471
* https://github.com/gradle/gradle/issues/8245
* https://github.com/gradle/gradle/issues/10585

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

aSemy commented on subprojects/distributions-dependencies/build.gradle.kts:
```diff
@@ -179,3 +179,33 @@ dependencies {
+
+/** `com.github.mwiede:jsch` is a drop-in replacement for `com.jcraft:jsch` */
+@CacheableRule
+object JschReplacementVersionRule : ComponentMetadataRule {
```
I had to copy this rule twice in

* `subprojects/distributions-dependencies/build.gradle.kts`
* `build-logic/build-platform/build.gradle.kts`

Is there a way to share it between all of the relevant java-platform plugins?

aSemy commented on gradle/dependency-management/capabilities.json:
```diff
@@ -102,5 +102,13 @@
         "selected": "cglib:cglib",
         "upgrade": "3.2.6"
+    },
+    {
+        "name": "jsch",
+        "providedBy": [
+            "com.jcraft:jsch", "com.github.mwiede:jsch"
+        ],
+        "selected": "com.github.mwiede:jsch",
+        "upgrade": "0.2.0"
```
This doesn't work as expected because I still see `com.jcraft:jsch` in the listed dependencies. Should `capabilities.json` apply to all projects, included builds, settings files, etc...?

![image](https://user-images.githubusercontent.com/897017/162003686-aee4980b-ebfd-41ed-a397-a3716c577fa7.png)

Is there a task in the Gradle project that will list the origin of all dependencies?

aSemy replied:
I've found an option in IntelliJ to show where a dependency is used, here's the report 

![image](https://user-images.githubusercontent.com/897017/163262557-361ad8f9-c150-48ff-8d43-3b42699d71e0.png)

I'll see if I can zap these remaining usages

aSemy commented:
From what I can tell this PR correctly replaces `com.jcraft:jsch` with `com.github.mwiede:jsch`

I ran a dependency report in IntelliJ and the output looks correct. The full file is 55MB so I won't paste it here!

But here's a screenshot to highlight some of the replacements, and the text of the screenshot in case you want to ctrl+f.

![image](https://user-images.githubusercontent.com/897017/162004597-ac0cd522-e160-47b6-b330-22e5b1f93273.png)

```xml
  <file path="$PROJECT_DIR$/subprojects/resources-sftp/src/main/java/org/gradle/internal/resource/transport/sftp/SftpClientFactory.java">
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/util/ArrayList.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/lang/String.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/Session.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/lang/Override.java" />
    <dependency path="$PROJECT_DIR$/subprojects/resources/src/main/java/org/gradle/api/resources/ResourceException.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/lang/Throwable.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/30.1.1-jre/fac882f35fd501581954dcaffa410ee4b5cd8d1b/guava-30.1.1-jre-sources.jar!/com/google/common/collect/ArrayListMultimap.java" />
    <dependency path="$PROJECT_DIR$/subprojects/base-services/src/main/java/org/gradle/internal/concurrent/CompositeStoppable.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/30.1.1-jre/fac882f35fd501581954dcaffa410ee4b5cd8d1b/guava-30.1.1-jre-sources.jar!/com/google/common/collect/Lists.java" />
    <dependency path="$PROJECT_DIR$/subprojects/core-api/src/main/java/org/gradle/api/credentials/PasswordCredentials.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/HostKey.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/net/URI.java" />
    <dependency path="$PROJECT_DIR$/subprojects/resources-sftp/src/main/java/org/gradle/internal/resource/transport/sftp/LockableSftpClient.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/ChannelSftp.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/JSch.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/30.1.1-jre/fac882f35fd501581954dcaffa410ee4b5cd8d1b/guava-30.1.1-jre-sources.jar!/com/google/common/collect/ListMultimap.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.30/71a1fdefa224da50f1db6306ce609981e20186c9/slf4j-api-1.7.30-sources.jar!/org/slf4j/LoggerFactory.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/util/List.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.30/71a1fdefa224da50f1db6306ce609981e20186c9/slf4j-api-1.7.30-sources.jar!/org/slf4j/Logger.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/30.1.1-jre/fac882f35fd501581954dcaffa410ee4b5cd8d1b/guava-30.1.1-jre-sources.jar!/com/google/common/collect/Multimap.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/UserInfo.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/Channel.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/lang/Object.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/JSchException.java" />
    <dependency path="$USER_HOME$/.jdks/temurin-11.0.14.1/lib/src.zip!/java.base/java/lang/UnsupportedOperationException.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.google.code.findbugs/jsr305/3.0.2/b19b5927c2c25b6c70f093767041e641ae0b1b35/jsr305-3.0.2-sources.jar!/javax/annotation/concurrent/ThreadSafe.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/HostKeyRepository.java" />
    <dependency path="$PROJECT_DIR$/subprojects/base-services/src/main/java/org/gradle/internal/concurrent/Stoppable.java" />
    <dependency path="$USER_HOME$/.gradle/caches/modules-2/files-2.1/com.github.mwiede/jsch/0.2.0/ba9e3ca6d111bcfc6b4178eb2a398584f9ce225f/jsch-0.2.0-sources.jar!/com/jcraft/jsch/Logger.java" />
    <dependency path="$PROJECT_DIR$/subprojects/resources-sftp/src/main/java/org/gradle/internal/resource/transport/sftp/SftpHost.java" />
  </file>
```

aSemy commented:
I've  just pushed another attempt. I'm still unable to completely exclude the outdated dependency. I've tried [following the documentation](https://docs.gradle.org/current/userguide/dependency_capability_conflict.html), but I still see the dependency appearing in the list of IntelliJ project dependencies. 

The project structure is really difficult to understand - it would be nice if there was a single place where dependency restrictions are defined. It seems like there's too many, and while I've tried to add the restriction in lots of places, I've missed some. Or maybe the 'ComponentMetadataRule' doesn't work. The change I made in `capabilities.json` didn't seem to work 🤷‍♀️

aSemy commented:
Haha I'm trying to fix the DCO problem but git wants me to manually review and merge 573 commits, one by one. No thank you. The DCO is pointless anyway right? It looks like it's the same as the committer/author field?

hegyibalint commented:
> Haha I'm trying to fix the DCO problem but git wants me to manually review and merge 573 commits, one by one. No thank you. The DCO is pointless anyway right? It looks like it's the same as the committer/author field?

Don't worry about it, after the review I will probably squash and sign the whole PR.

hegyibalint commented:
@bot-gradle test RFM

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"84f7b5f2ef3eb4d9a96c7dae3b7d6d108244b51d","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=51372336)

ljacomet commented on build-logic/build-platform/build.gradle.kts:
```diff
@@ -73,5 +73,45 @@ dependencies {
         }
         api("net.bytebuddy:byte-buddy") { version { strictly("1.10.21") } }
         api("org.objenesis:objenesis") { version { strictly("3.1") } }
+        api("com.github.mwiede:jsch:0.2.0") {
```
I do no think this is needed. This is because `jsch` is not a direct dependency for the build logic. Though it comes transitively from `jgit`

aSemy replied:
I'm not sure if these files count as 'build logic', but jsch is added as a dependency here:

* https://github.com/gradle/gradle/blob/36cad0c3425102540db84b80ec85702f69b948a0/subprojects/internal-performance-testing/build.gradle.kts#L61
* https://github.com/gradle/gradle/blob/36cad0c3425102540db84b80ec85702f69b948a0/subprojects/version-control/build.gradle.kts#L22
* https://github.com/gradle/gradle/blob/36cad0c3425102540db84b80ec85702f69b948a0/subprojects/version-control/build.gradle.kts#L35
* https://github.com/gradle/gradle/blob/36cad0c3425102540db84b80ec85702f69b948a0/subprojects/resources-sftp/build.gradle.kts#L15
* https://github.com/gradle/gradle/blob/36cad0c3425102540db84b80ec85702f69b948a0/subprojects/internal-integ-testing/build.gradle.kts#L90


And it's being used directly in code https://github.com/gradle/gradle/blob/36cad0c3425102540db84b80ec85702f69b948a0/subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java#L20

ljacomet commented on build-logic/build-platform/build.gradle.kts:
```diff
@@ -73,5 +73,45 @@ dependencies {
+        }
+    }
+
+    components {
```
None of the changes below are required. There is no real dependency resolution happening in a `java-platform` project.

aSemy replied:
https://github.com/gradle/gradle/pull/20387#discussion_r861992460

ljacomet commented on build-logic/dependency-modules/src/main/kotlin/gradlebuild.dependency-modules.gradle.kts:
```diff
@@ -70,6 +70,8 @@ dependencies {
             setOf("gradle-tooling-api")
         )
+
+        all<JschReplacementVersionRule>()
```
The change needed here is closer to what `ReplaceCglibNodepWithCglibRule` does and should be applied to `jgit` only.

ljacomet commented on gradle/dependency-management/capabilities.json:
```diff
@@ -102,5 +102,13 @@
         "upgrade": "3.2.6"
+    },
+    {
+        "name": "jsch",
```
This is useful to make sure we do not accidentally re-introduce the old `jsch` in the future. But on its own, it will not solve all issues as you discovered and commented.

aSemy replied:
https://github.com/gradle/gradle/pull/20387#discussion_r861992460

ljacomet commented on subprojects/distributions-dependencies/build.gradle.kts:
```diff
@@ -178,4 +178,41 @@ dependencies {
         api(libs.xmlunit)               { version { strictly("1.6") }}
     }
+
+    components {
```
I do not think this is required either. The replacement and conflict detection should be handled elsewhere.

aSemy replied:
okay, I removed the 'jsch replacement' rule (and in other places)

martin-braun commented:
I do not understand the relationship between `gradle`, `jsch` and `jgit` in its entirety, but a senior member of the JGit forum says: 

> If gradle still uses JSch with JGit it is using no longer supported JGit code.

Source: https://www.eclipse.org/forums/index.php?t=msg&th=1110330&goto=1851614&#msg_1851614

aSemy commented:
@martin-braun That's really useful info! Gradle has got a hard dependency on jsch, so the better solution to this problem would be to migrate from jsch to the Apache MINA replacement. I'll see if I can 

---

@ljacomet Thanks for reviewing! I don't really understand what's going on though...

> I believe some issues in solving it come from a misunderstanding of leveraging capability conflicts as the only solution for replacing the old `jsch` with the new one.

If it's a misunderstanding, then it's a documented misunderstanding. I followed the instructions here:

https://docs.gradle.org/current/userguide/dependency_capability_conflict.html

Aside: The transitive dependencies docs could be improved by focusing less on teaching or explaining how Gradle solves things, and giving shorter "if you want to do X, then use <code>" instructions. 

I don't understand what `ReplaceCglibNodepWithCglibRule` is supposed to do, and why it's better or worse than the `JschReplacementVersionRule` that I attempted... but anyway, based on @martin-braun's comment I think the better solution is to upgrade jgit to a newer version.

aSemy commented on subprojects/base-services/src/main/java/org/gradle/internal/FileUtils.java:
```diff
@@ -97,7 +97,7 @@ public static File assertInWindowsPathLengthLimitation(File file) {
      * This method does not access the file system.
      * It is agnostic to whether a given file object represents a regular file, directory or does not exist.
-     * That is, the term “file” is used in the java.io.File sense, not the regular file sense.
+     * That is, the term "file" is used in the java.io.File sense, not the regular file sense.
```
See https://github.com/gradle/gradle/issues/20612

aSemy commented on subprojects/version-control/src/main/java/org/gradle/vcs/git/internal/GitVersionControlSystem.java:
```diff
@@ -194,16 +188,24 @@ private static GradleException wrapGitCommandException(String commandName, URI r
-                    }
-                });
+
+                SshdSessionFactory factory = new SshdSessionFactory(
+                    new JGitKeyCache(),
+                    new DefaultProxyDataFactory()
+                );
+
+                try {
+                    Runtime.getRuntime().addShutdownHook(new Thread(factory::close));
+                } catch (IllegalStateException e) {
+                    // ignore - the VM is already shutting down
+                }
+
+                sshTransport.setSshSessionFactory(factory);
```
It seems this is how to replace jsch

https://www.eclipse.org/forums/index.php?t=msg&th=1106995&goto=1838158&#msg_1838158

https://git.eclipse.org/r/plugins/gitiles/jgit/jgit/+/refs/heads/master/org.eclipse.jgit.pgm/src/org/eclipse/jgit/pgm/TextBuiltin.java#219

aSemy commented:
Okay, I think the task has been done! I can't see jsch anymore!

![image](https://user-images.githubusercontent.com/897017/166066977-f1a2fa8e-7366-4aa8-b743-17aca7cc13b8.png)


In summary

* I updated jgit to version 6 (I had to do it in two places)
  * https://github.com/gradle/gradle/pull/20387/files#diff-28c8d424ea5c745ef9d0e62b9060f05fd7765ca806a1df5b721d95c377b2e229L62-R62
  * https://github.com/gradle/gradle/pull/20387/files#diff-961415f8ddcdd04d65d758b9f9a23809f36a53eb936e7e062129c74f1c23d61cL100
* added jgit.ssh.apache, which is some sort of replacement for jsch
   https://github.com/gradle/gradle/pull/20387/files#diff-961415f8ddcdd04d65d758b9f9a23809f36a53eb936e7e062129c74f1c23d61cR101
* updated the jgit code to use jgit.ssh.apache
   https://github.com/gradle/gradle/pull/20387/files#r862114501
   This seems to mean that the default branch is 'main' instead of 'master' now, so I updated the tests.

martin-braun commented:
> This pull request has been automatically marked as stale because it has not had recent activity. Given the limited bandwidth of the team, it will be closed if no further activity occurs. If you intend to work on this pull request, please ask the team to reopen the PR or push a new PR. Thank you for your contributions.

Bad bot, don't!

martin-braun commented:
Good bot! :)

alllex commented:
GitHub shows that this PR contains 10k commits. I think it's perhaps best to cleanly rebase the branch on top of the latest `master`.

wdroste commented:
Just an FYI, if you do the recommended `gradle` builds the use SSH for tagging etc basically releases are broken.

https://bitbucket.org/blog/ssh-host-key-changes

ljacomet commented:
Thank you for this attempt. It turned out to be a bit more involved but is now ready to be merged. Closing in favor of the new PR handling this:
* #27777
